Code/GraphMol/Bond.cpp

```diff
@@ -322,48 +322,51 @@ const Atom *findHighestCIPNeighbor(const Atom *atom, const Atom *skipAtom) {
   return bestCipRankedAtom;
 }
 
-INT_VECT *findStereoAtoms(const Bond *bond) {
-  if (!bond->hasOwningMol()) {
-    return new INT_VECT;
-  }
-  switch (bond->getStereo()) {
-    case Bond::BondStereo::STEREOE:
-    case Bond::BondStereo::STEREOZ: {
-      const Atom *startStereoAtom =
-          findHighestCIPNeighbor(bond->getBeginAtom(), bond->getEndAtom());
-      const Atom *endStereoAtom =
-          findHighestCIPNeighbor(bond->getEndAtom(), bond->getBeginAtom());
-
-      if (startStereoAtom != nullptr && endStereoAtom != nullptr) {
-        int startStereoAtomIdx = static_cast<int>(startStereoAtom->getIdx());
-        int endStereoAtomIdx = static_cast<int>(endStereoAtom->getIdx());
-
-        return new INT_VECT({startStereoAtomIdx, endStereoAtomIdx});
-
-        break;
-      }
+void findStereoAtoms(const Bond *bond, INT_VECT *&p_stAt) {
+  if (bond->hasOwningMol() &&
+      (bond->getStereo() == Bond::BondStereo::STEREOE ||
+       bond->getStereo() == Bond::BondStereo::STEREOZ)) {
+    const Atom *startStereoAtom =
+        findHighestCIPNeighbor(bond->getBeginAtom(), bond->getEndAtom());
+    const Atom *endStereoAtom =
+        findHighestCIPNeighbor(bond->getEndAtom(), bond->getBeginAtom());
+
+    if (startStereoAtom != nullptr && endStereoAtom != nullptr) {
+      int startStereoAtomIdx = static_cast<int>(startStereoAtom->getIdx());
+      int endStereoAtomIdx = static_cast<int>(endStereoAtom->getIdx());
+
+      p_stAt = new INT_VECT({startStereoAtomIdx, endStereoAtomIdx});
+      return;
     }
-    default:
-      return new INT_VECT;
   }
+  p_stAt = new INT_VECT;
 }
 
 }  // namespace
 
 const INT_VECT &Bond::getStereoAtoms() const {
   if (!dp_stereoAtoms) {
-    const_cast<Bond *>(this)->dp_stereoAtoms = findStereoAtoms(this);
+    Bond *mutableSelf = const_cast<Bond *>(this);
+#ifdef RDK_THREADSAFE_SSS
+    std::call_once(mutableSelf->d_stAtMutex, findStereoAtoms, mutableSelf,
+                   mutableSelf->dp_stereoAtoms);
+#else
+    findStereoAtoms(this, mutableSelf->dp_stereoAtoms);
+#endif
   }
   return *dp_stereoAtoms;
 };
 
 INT_VECT &Bond::getStereoAtoms() {
   if (!dp_stereoAtoms) {
-    dp_stereoAtoms = findStereoAtoms(this);
+#ifdef RDK_THREADSAFE_SSS
```

Comment:
I don't think this is the right solution.  It still has the issue of casting away const, and I feel the call_once is kind of odd here.  I think the right solution is to either set this when assigning stereochemistry, or perhaps move the calls to bond::updatePropertyCache, although I'm unsure if this is called in the correct places.

However, there may be a new bug introduced here:  I'm troubled by the fact that findStereoAtoms apparently can't be called more than once.  What if the molecule is edited?  Will findStereoAtoms every be called again?  I don't think it will, and there is no way to reset dp_stereoAtoms.

The previous code was "only" creating the INT_VECT when it didn't exist, it wasn't actually doing anything with it so I don't think we saw this issue.

Reply:
Is there any chance we can punt this code into addReactantTemplate and addProductTemplate for now?

```
void setStereoAtoms(Bond *bond) {
  if (bond->hasOwningMol() &&
      (bond->getStereo() == Bond::BondStereo::STEREOE ||
       bond->getStereo() == Bond::BondStereo::STEREOZ)) {
    const Atom *startStereoAtom =
        findHighestCIPNeighbor(bond->getBeginAtom(), bond->getEndAtom());
    const Atom *endStereoAtom =
        findHighestCIPNeighbor(bond->getEndAtom(), bond->getBeginAtom());

     if (startStereoAtom != nullptr && endStereoAtom != nullptr) {
      int startStereoAtomIdx = static_cast<int>(startStereoAtom->getIdx());
      int endStereoAtomIdx = static_cast<int>(endStereoAtom->getIdx());

       bond->setStereoAtoms(startStereoAtomIdx, endStereoAtomIdx);
    }
  }
}

unsigned int addReactantTemplate(ROMOL_SPTR mol) {
    this->df_needsInit = true;
    for(auto bond: mol->bonds()) {
       setStereoAtoms(bond);
   }
    this->m_reactantTemplates.push_back(mol);
    return rdcast<unsigned int>(this->m_reactantTemplates.size());
  }
```

Reply:
addReactantTemplate has the benefit of REALLY not being thread safe :)

Reply:
I have no idea how I managed, but I did read the doc for `call_once()` exactly the opposite as it is, and somehow got the idea that `once_flag` did reset after the execution.

Anyway, moving these changes to `addReactantTemplate()` seems like a good idea.